addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go

```diff
@@ -0,0 +1,238 @@
+// Copyright YEAR VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/utils/pointer"
+
+	clusterapipatchutil "sigs.k8s.io/cluster-api/util/patch"
+
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addons/v1alpha1"
+)
+
+// SecretGenControllerConfigReconciler reconciles a SecretGenControllerConfig object
+type SecretGenControllerConfigReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//SecretGenControllerConfigSpecYaml is used for yaml marshalling for secretGen controller config
+type SecretGenControllerConfigSpecYaml struct {
+	Namespace       string `yaml:"namespace,omitempty"`
+	CreateNamespace bool   `yaml:"createNamespace,omitempty"`
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=secretgencontrollerconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=secretgencontrollerconfigs/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *SecretGenControllerConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("secretgencontrollerconfig", req.NamespacedName)
+
+	// get secret-gen config object
+	secretGenConfig := &addonsv1alpha1.SecretGenControllerConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secretGenConfig); err != nil {
+		log.Error(err, "unable to fetch SecretGenControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// verify that the cluster related to config is present
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
```

Comment:
Does the ```req.NamespacedName``` really involve cluster name and namespace? Did you test this on an existing cluster?

Reply:
Question: Using `req.NamespacedName` for both `xxxConfig` and `Cluster` implicitly assumes two objs always have the same name and under the same namespace. We need to be very careful even we are opinionated on naming `xxxConfig`.